common/persistence/history_branch_util.go

```diff
@@ -0,0 +1,91 @@
+// The MIT License
+//
+// Copyright (c) 2020 Temporal Technologies Inc.  All rights reserved.
+//
+// Copyright (c) 2020 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	enumspb "go.temporal.io/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/persistence/serialization"
+	"go.temporal.io/server/common/primitives"
+)
+
+type (
+	HistoryBranchUtil interface {
+		// NewHistoryBranchToken initializes a new history branch token
+		NewHistoryBranchToken(treeID string, branchID *string, ancestors []*persistencespb.HistoryBranchRange) ([]byte, error)
+		// ParseHistoryBranchInfo parses the history branch for branch information
+		ParseHistoryBranchInfo(branchToken []byte) (*persistencespb.HistoryBranch, error)
+		// UpdateHistoryBranchInfo updates the history branch with branch information
+		UpdateHistoryBranchInfo(branchToken []byte, branchInfo *persistencespb.HistoryBranch) ([]byte, error)
+	}
+
+	historyBranchUtilImpl struct {
+	}
+)
+
+// NewHistoryBranchUtil returns a history branch utility
+func NewHistoryBranchUtil() HistoryBranchUtil {
+	return &historyBranchUtilImpl{}
+}
```

Comment:
I like moving these functions to a separate utility file. However, since this object has no state and no I/O operations, the code would be more concise and transparent if we used functions here instead of object methods.

Reply:
This is an extension point for injecting different logic.